torch/_library/autograd.py

```diff
@@ -31,8 +34,136 @@ def backward(ctx, *grads):
+        nonlocal output_spec
+        nonlocal result_is_tuple
+        result_is_tuple = isinstance(result, tuple)
+        if not result_is_tuple:
+            result = (result,)
+        nonlocal output_spec
+        flat_result, output_spec = flatten(result, not_list_of_tensor)
+
+        # Save the input_spec/output_spec for backward because another call to
+        # .apply will override the nonlocals.
+        if hasattr(ctx, "_pt_metadata"):
+            raise RuntimeError(
+                "Please don't set ctx._pt_metadata; PyTorch uses it to store info"
+            )
+        ctx._pt_metadata = (input_spec, output_spec)
+
+        return tuple(flat_result)
+
+    def new_backward(ctx, *grads):
+        if not hasattr(ctx, "_pt_metadata"):
+            raise NotImplementedError(
+                "NYI: calling supports_tensorlist autograd.Function.backward directly. "
+                "This will automatically get called by PyTorch autograd. "
+                "Please file an issue if you need this."
+            )
+
+        input_spec, output_spec = ctx._pt_metadata
+        grads = unflatten(list(grads), output_spec)
+        grad_inputs = orig_backward(ctx, *grads)
+        if not isinstance(grad_inputs, tuple):
+            grad_inputs = (grad_inputs,)
+        # Assume that any Nones in the backward are Tensors.
+        # If the forward has an arg that is [1, 2, 3], the backward should
+        # return None as the grad.
+        # If the forward has an arg that is [tensor, tensor], the backward
+        # may return [None, None], [grad, None], [None, grad], or [grad, grad].
+        flat_grad_inputs, grad_inputs_spec = flatten(
+            grad_inputs, not_list_of_optional_tensor
+        )
+        if grad_inputs_spec != input_spec:
+            raise RuntimeError(
+                f"Expected the return from backward to be of the same structure "
+                f"as the inputs. Got: {grad_inputs_spec} (return from backward), "
+                f"{input_spec} (inputs)"
+            )
+        return tuple(flat_grad_inputs)
+
+    def new_apply(*args):
+        nonlocal input_spec
+        if input_spec is not None:
+            raise NotImplementedError(
+                "NYI: Recursive call to autograd.Function decorated with "
+                "`supports_tensorlist`. Please file an issue."
+            )
+        try:
+            flat_args, input_spec = flatten(args, is_leaf=not_list_of_tensor)
+            result = orig_apply(*flat_args)  # type: ignore[misc]
+        finally:
+            input_spec = None
+        assert output_spec is not None
+        result = unflatten(list(result), output_spec)
+        if not result_is_tuple:
+            assert isinstance(result, tuple)
+            assert len(result) == 1
+            return result[0]
+        return result
+
+    cls.forward = new_forward
+    cls.backward = new_backward
+    cls.apply = new_apply
+    return cls
+
+
+def not_list_of_tensor(tree):
```

Comment:
I don't recall the full details, but what's the link between our TensorList and tuple(Tensor,...) ?
I guess we convert tuple(Tensor, ...) to TensorList on the way in. But do we return actual lists?
I'm ok with being very strict here but that might be surprising to some users.

Reply:
For inputs, the pytorch dispatcher converts them to list
For outputs, the pytorch dispatcher converts tuples to list

I expect users to have trouble with the strictness (there are other problems, like not being able to pass None as the gradient for a TensorList), but I don't want to relax them in this PR (it's already getting complicated).